src/stats.c

```diff
@@ -286,90 +328,82 @@ stats_arena_bins_print(void (*write_cb)(void *, const char *), void *cbopaque,
 		CTL_M2_M4_GET("stats.arenas.0.bins.0.curslabs", i, j, &curslabs,
 		    size_t);
 
-		if (json) {
-			malloc_cprintf(write_cb, cbopaque,
-			    "\t\t\t\t\t{\n"
-			    "\t\t\t\t\t\t\"nmalloc\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"ndalloc\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"curregs\": %zu,\n"
-			    "\t\t\t\t\t\t\"nrequests\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"nfills\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"nflushes\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"nreslabs\": %"FMTu64",\n"
-			    "\t\t\t\t\t\t\"curslabs\": %zu%s\n",
-			    nmalloc, ndalloc, curregs, nrequests, nfills,
-			    nflushes, nreslabs, curslabs, mutex ? "," : "");
-			if (mutex) {
-				uint64_t mutex_stats_64[mutex_prof_num_uint64_t_counters];
-				uint32_t mutex_stats_32[mutex_prof_num_uint32_t_counters];
-				read_arena_bin_mutex_stats(i, j, mutex_stats_64, mutex_stats_32);
-				mutex_stats_output_json(write_cb, cbopaque,
-				    "mutex", mutex_stats_64, mutex_stats_32, "\t\t\t\t\t\t", true);
-			}
-			malloc_cprintf(write_cb, cbopaque,
-			    "\t\t\t\t\t}%s\n",
-			    (j + 1 < nbins) ? "," : "");
-		} else if (!in_gap) {
-			size_t availregs = nregs * curslabs;
-			char util[6];
-			if (get_rate_str((uint64_t)curregs, (uint64_t)availregs,
-			    util)) {
-				if (availregs == 0) {
-					malloc_snprintf(util, sizeof(util),
-					    "1");
-				} else if (curregs > availregs) {
-					/*
-					 * Race detected: the counters were read
-					 * in separate mallctl calls and
-					 * concurrent operations happened in
-					 * between. In this case no meaningful
-					 * utilization can be computed.
-					 */
-					malloc_snprintf(util, sizeof(util),
-					    " race");
-				} else {
-					not_reached();
-				}
-			}
-			uint64_t mutex_stats_64[mutex_prof_num_uint64_t_counters];
-			uint32_t mutex_stats_32[mutex_prof_num_uint32_t_counters];
-			if (mutex) {
-				read_arena_bin_mutex_stats(i, j, mutex_stats_64, mutex_stats_32);
-			}
+		if (mutex) {
```

Comment:
Should this be combined with the branch at L353?

Reply:
I sort of like the separation of having all the mallctlls happening in one place, followed by all the emitting happening in one place (as opposed to mallctl -> json emitting -> maybe some more mallctl -> table emitting). WDYT?

Reply:
That sounds good too. Keep it as is then.